chat-bots/src/Data/Chat/Bot/Monoidal.hs

```diff
@@ -97,5 +87,4 @@ infixr 9 /.\
 infixr 9 \/
 
 (\/) :: Monad m => Bot m s i o -> Bot m s i' o' -> Bot m s (i \/ i') (o \/ o')
-(\/) (Bot b1) (Bot b2) =
-  Bot $ \s -> either (fmap (first Left) . b1 s) (fmap (first Right) . b2 s)
+(\/) b1 b2 = invmapBot fst split' $ Trifunctor.combine (b1, b2)
```

Comment:
Here i'm using `combine @-> @(,) @Either @Either @(,)` which gives you a product of two state types that happen to be the same. I use `invmapBot` to collapse the product. Notice there is a bias to the first projection due to this. Does that matter? @masaeedu

Reply:
Hmm. It's not at least trivially clear to me that this is the same function, I need to think about it a bit more.

That said, just uncurrying `combine @-> @(,) @Either @Either @(,)` gives a perfectly good infix operator of type:

```haskell
Bot m s i o -> Bot m s' i' o' -> Bot m (s /\ s') (i \/ i') (o \/ o')
```

Would it cause too much breakage downstream to generalize the type?

As far as recovering exactly our existing operation from `Trifunctor.combine`, the additional challenge is that thinking of `Bot` as a monoidal functor in `s` forces us to grapple with invariant monoidal functors.

While we have a basic treatment of these (we just treat them like plain old monoidal functors out of `Set^op x Set`), we need to think about co/end calculus and _di_-natural transformations to deal with these well. I don't really have a good intuition for these yet (although they've been on my goals for a while), nor do we yet account for them in the library, so it's a bit of an outstanding project. Here for example we want to work with the end of a (monoidal) profunctor into a category of monoidal bifunctors.

As a concrete step it might help to abstract away from all this bot stuff a bit somewhere and more generally investigate the monoidal-functor-hood of coalgebras of monoidal functors.

Reply:
> Would it cause too much breakage downstream to generalize the type?

I don't think we actually use `/.\` in the library so no it wouldn't be problem. 

> As far as recovering exactly our existing operation from `Trifunctor.combine,` the additional challenge is that thinking of `Bot` as a monoidal functor in `s` forces us to grapple with invariant monoidal functors.

I agree and I think also conceptually the `combine @-> @(,) @Either @Either @(,)` instance makes more sense for `Bot` then an invariant `s`. I might be wrong but with an invariant `s` wont the two `Bots` we are tensoring either need to either have the same `s` or not use `s`?

> While we have a basic treatment of these (we just treat them like plain old monoidal functors out of Set^op x Set), we need to think about co/end calculus and di-natural transformations to deal with these well. I don't really have a good intuition for these yet (although they've been on my goals for a while), nor do we yet account for them in the library, so it's a bit of an outstanding project. Here for example we want to work with the end of a (monoidal) profunctor into a category of monoidal bifunctors.

I'd love to get on a video call with you at some point to discuss and learn more about this.

> As a concrete step it might help to abstract away from all this bot stuff a bit somewhere and more generally investigate the monoidal-functor-hood of coalgebras of monoidal functors.

Is this something that woud live in `monoidal-functors`? Again, I would love to get on a call to discuss this.

Reply:
Since we aren't actually using `/.\` at the moment, I am going to redefine it as `combine @-> @(,) @Either @Either @(,)`.